iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/vector.inl

```diff
@@ -249,26 +254,33 @@ inline bool vector<T, Capacity>::resize(const uint64_t count, const Targs&... ar
 template <typename T, uint64_t Capacity>
 inline T* vector<T, Capacity>::data() noexcept
 {
+    // AXIVION Next Construct AutosarC++19_03-A5.2.4 : Type-safety ensured by template parameter
+    // NOLINTNEXTLINE (cppcoreguidelines-pro-type-reinterpret-cast)
```

Comment:
We should use here:
```cxx
return &at(0);
```
Then we do not need to add an axivion and clang-tidy exception.

Reply:
Here I would keep the current code. `&at(0)` will change the semantics since it will terminate when the vector is empty. If this is desired, then it is a different question but with the current semantics it breaks the code. Since there is also `front`, I would keep the current semantic of `data`

Reply:
@elBoberido @elfenpiff  would vote for changing the semantics. I can add a note to the release notes. Anyway for now I used `at_unchecked` and let the non-`const` `data()` call the `const` one.

Reply:
@mossmaurice this is a good question and I might take counsel with my pillow first.

In general we should try to stay consistent with other constructs, e.g. `string::c_str` which is the equivalent to this `data` function.

Something to keep in mind is also what is dangerous in this situation. Getting the pointer is not dangerous since the memory is there. An out of bounds access is dangerous but that can also happen when the vector contains e.g. 1 item. Therefore the tendency is to keep the current semantic but lets see what the pillow counsel reveals